tests/Type/Doctrine/Query/QueryResultTypeWalkerTestCase.php

```diff
@@ -0,0 +1,349 @@
+use function array_shift;
+use function assert;
+use function class_exists;
+use function count;
+use function property_exists;
+use function sprintf;
+use function version_compare;
+use const PHP_VERSION_ID;
+
+abstract class QueryResultTypeWalkerTestCase extends PHPStanTestCase
+{
+
+	/** @var EntityManagerInterface */
+	protected static $em;
+
+	/** @var DescriptorRegistry */
+	private $descriptorRegistry;
+
+	abstract protected static function getEntityManagerPath(): string;
+
+	public static function getAdditionalConfigFiles(): array
+	{
+		return [
+			__DIR__ . '/../data/QueryResult/config.neon',
+		];
+	}
+
+	public static function setUpBeforeClass(): void
+	{
+		$em = require static::getEntityManagerPath();
+		self::$em = $em;
+
+		$schemaTool = new SchemaTool($em);
+		$classes = $em->getMetadataFactory()->getAllMetadata();
+		$schemaTool->createSchema($classes);
+
+		$dataOne = [
+			'intColumn' => [1, 2],
+			'stringColumn' => ['A', 'B'],
+			'stringNullColumn' => ['A', null],
+		];
+
+		$dataMany = [
+			'intColumn' => [1, 2],
+			'stringColumn' => ['A', 'B'],
+			'stringNullColumn' => ['A', null],
+		];
+
+		$dataJoinedInheritance = [
+			'parentColumn' => [1, 2],
+			'parentNullColumn' => [1, null],
+			'childColumn' => [1, 2],
+			'childNullColumn' => [1, null],
+		];
+
+		$dataSingleTableInheritance = [
+			'parentColumn' => [1, 2],
+			'parentNullColumn' => [1, null],
+			'childNullColumn' => [1, null],
+		];
+
+		$id = 1;
+
+		foreach (self::combinations($dataOne) as $combination) {
+			[$intColumn, $stringColumn, $stringNullColumn] = $combination;
+			$one = new One();
+			$one->id = (string) $id++;
+			$one->intColumn = $intColumn;
+			$one->stringColumn = $stringColumn;
+			$one->stringNullColumn = $stringNullColumn;
+			$embedded = new Embedded();
+			$embedded->intColumn = $intColumn;
+			$embedded->stringColumn = $stringColumn;
+			$embedded->stringNullColumn = $stringNullColumn;
+			$nestedEmbedded = new NestedEmbedded();
+			$nestedEmbedded->intColumn = $intColumn;
+			$nestedEmbedded->stringColumn = $stringColumn;
+			$nestedEmbedded->stringNullColumn = $stringNullColumn;
+			$embedded->nestedEmbedded = $nestedEmbedded;
+			$one->embedded = $embedded;
+			$one->manies = new ArrayCollection();
+
+			foreach (self::combinations($dataMany) as $combinationMany) {
+				[$intColumnMany, $stringColumnMany, $stringNullColumnMany] = $combinationMany;
+				$many = new Many();
+				$many->id = (string) $id++;
+				$many->intColumn = $intColumnMany;
+				$many->stringColumn = $stringColumnMany;
+				$many->stringNullColumn = $stringNullColumnMany;
+				$many->datetimeColumn = new DateTime('2001-01-01 00:00:00');
+				$many->datetimeImmutableColumn = new DateTimeImmutable('2001-01-01 00:00:00');
+				$many->simpleArrayColumn = ['foo'];
+				$many->one = $one;
+				$one->manies->add($many);
+				$em->persist($many);
+			}
+
+			$em->persist($one);
+		}
+
+		foreach (self::combinations($dataJoinedInheritance) as $combination) {
```

Comment:
I'm not sure we need this complexity. I ended with super simple test fetching a single expression and verifying that type in https://github.com/janedbal/php-database-drivers-fetch-test All we need here is to check that against the inferred PHPStan type.

And I feel this is the best approach to accomplish what we need as it can be [quite readable in the test itself](https://github.com/janedbal/php-database-drivers-fetch-test/blob/master/tests/PhpDatabaseDriverTest.php#L127)

Reply:
I just copied the existing `QueryResultTypeWalkerTest` with the whole dataprovider.

Do you think we should split it ?
- Only single expression for platforms tests
- single expression and complex expression for PDO tests in order to validate PHPStan for these